File: go/problem_solving/leet_code/linked_list/reverse.go
package linked_list

//https://leetcode.com/problems/reverse-linked-list/

type ListNode struct {
	Val  int
	Next *ListNode
}

func reverseList(head *ListNode) *ListNode {
	var prev *ListNode
	for head != nil {
		next := head.Next
		head.Next = prev
		prev = head
		head = next
	}
	return prev
}

func reverseListRecursive(head *ListNode) *ListNode {
	if head == nil {
		return nil
	}
	newHead := head
	if head.Next != nil {
		newHead = reverseListRecursive(head.Next)
		//Reverse link between next node and head
		head.Next.Next = head
	}
	head.Next = nil
	return newHead
}

File: algorithms-data-structures/ts/data-structures/linked-list/single/reverse/reverse.ts
import {ListNode} from "../list-node";

export class Solution {
  reverseList(head: ListNode | null): ListNode | null {
    let prev = null;
    let curr = head;
    while (curr !== null) {
      let nextTemp = curr.next;
      curr.next = prev;
      prev = curr;
      curr = nextTemp;
    }
    return prev;
  }
}
